Valigator.Core/Core/DataDescriptor.cs

```diff
@@ -21,5 +21,20 @@ public DataDescriptor(Type propertyType, IStateDescriptor stateDescriptor, IValu
 			StateDescriptor = stateDescriptor ?? throw new ArgumentNullException(nameof(stateDescriptor));
 			ValueDescriptors = valueDescriptors ?? throw new ArgumentNullException(nameof(valueDescriptors));
 		}
+
+		public override bool Equals(object obj)
+			=> obj is DataDescriptor descriptor && Equals(descriptor);
+
+		public bool Equals(DataDescriptor other)
+			=> EqualityComparer<Type>.Default.Equals(PropertyType, other.PropertyType) && EqualityComparer<IStateDescriptor>.Default.Equals(StateDescriptor, other.StateDescriptor) && ValueDescriptors.SequenceEqual(other.ValueDescriptors);;
+
+		public override int GetHashCode()
+		{
+			var hashCode = 1881245575;
+			hashCode = hashCode * -1521134295 + EqualityComparer<Type>.Default.GetHashCode(PropertyType);
+			hashCode = hashCode * -1521134295 + EqualityComparer<IStateDescriptor>.Default.GetHashCode(StateDescriptor);
+			hashCode = hashCode * -1521134295 + EqualityComparer<IReadOnlyList<IValueDescriptor>>.Default.GetHashCode(ValueDescriptors);
```

Comment:
This hashcode won't work. You'll need to aggregate the hashcodes of all items inside the list. You should probably use (current, item) => current * -1521134295 + item.GetHashCode() or something like that.